cli/lock.ml

```diff
@@ -309,6 +359,18 @@ let local_packages ~versions repo =
   let open Result.O in
   Project.all_local_packages repo >>| package_version_map ~versions
 
+let extract_source_config ~opam_monorepo_cwd ~opam_files target_packages =
+  let open Result.O in
+  let target_opam_files =
+    List.map (OpamPackage.Name.Set.elements target_packages) ~f:(fun name ->
+        snd (OpamPackage.Name.Map.find name opam_files))
```

Comment:
In terms of performances, I think this will never manipulate sets of packages large enough for this to matter.

`find` is Map.find so it is `O(log(n))` and so is `Set.mem`. I think it would be strictly equivalent in term of complexity. Assuming `N` is the number of local packages and `n` the number of target packages (which is a subset of local packages btw), this is either `n.log.(N)` or `N.log(n)` and I think the former is best here.

Indeed `find` can raise but at this point, we know it won't and if it does, we messed up somewhere so it makes sense to raise, rather than report an error to the user. If I was to use `find_opt` I would add an `assert false` in the `None` branch.

Reply:
I agree that we probably won't reach a territory where the `n` is big enough to matter. Yet I still often find more efficient code kinda nicer to read because it also tends to do less.

You're right, a sensibly implemented `Map.find` should be `O(log n)` but `Set.mem` should be `O(1)` still (somewhat unexpectedly, the OCaml documentation does not actually specify the big-O notation of these operations — a shame).

In general I find that `target_opam_files` seems more of a transformation of `opam_files` than `target_packages`: it filters it down by checking whether they're targets, not as a transformation of `target_packages` to look each of them up in the `opam_files` map. This is, admittedly, personal taste, but I find this a bit easier to understand:

```ocaml
let target_opam_files = OpamPackage.Name.Map.fold (fun name (_, opam_file) target_opam_files ->
  match OpamPackage.Name.Set.mem name target_packages with
  | true -> opam_file::target_opam_files
  | false -> target_opam_files) opam_files []
```

YMMV of course.

Reply:
Why would set lookup be O(1) ? That sounds strictly impossible to me. You must be thinking about hash tables, sets are implemented as binary trees in OCaml, same as maps. Looking up in a map or a set should be identical in complexity if they share the same base data structure.

Either way, this is I think only a matter of taste indeed. What I like in my solution is that it doesn't include a `fold` which are in my opinion harder to read as you have to guess what data structure they are building by reading the code. It uses a map,making it clear that we get one opam file for each target package.

if that's fine by you, let's stick to the current implementation, we can always refine it later if we find something better.

Reply:
It is not OCaml but this [list of execution times](https://gist.github.com/FedericoPonzi/8d5094dbae33cbb94536a73f62d1c1a0) in Java mentions that `contains` is `O(1)` which to me is very much a reason to use a set datastructure to begin with - fast membership tests.

Although, I got curious and you're right, the documentation of OCaml states this:

> The implementation uses balanced binary trees, and is therefore reasonably efficient: insertion and membership take time logarithmic in the size of the set, for instance.

Ooof! This basically makes sets just unit maps and not particularly useful on their own. Disappointing.

Reply:
Note that we're not using sets for membership tests but rather to enforce uniqueness of members at the type level here.

To be honest, on this set size, I think tree and hash table based sets would have similar performances. I understand the disappointment but when we need better perfs, we can use better suited data structures. It's not really necessary here!